testdata/p4_16_samples/virtual3.p4

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include <core.p4>
+
+// In arch file
+extern Virtual {
+    Virtual();
+    // abstract methods must be implemented
+    // by the users
+    void increment();
+    @synchronous(increment) abstract bit<16> f(in bit<16> ix);
```

Comment:
So the meaning of this is that increment calls f?
Shouldn't the annotation be on increment instead?

Reply:
Yes, the annotation means that increment might call f (but is not guarenteed to).  At the moment I have the annotations on the abstract methods, listing the other methods that might call them, but it could equivalently be done the opposite way (put @might_call annotations on methods listing the other methods it might call?)

Reply:
If we want to support the case where the same method may be called both synchronously and asynchronously it may make sense to put the annotation on the caller.

I propose these rules:
- a method (virtual or non-virtual) that may be called asynchronously *must* be annotated with `@async`. This means that the method may be called at any time, from any thread. Such methods cannot access any P4 locals, only their own parameters. Such methods can also be called synchronously.
- a non-virtual method that may call `callee` is labeled with `@calls(callee)`

With these rules we have that asynchronous methods are always live, but can never reach into locals. We should always run def-use for them. Synchronous methods may use locals, but we should know all their callers, so if they are unreachable they are actually dead. Also, it is an error for an asynchronous method to call a synchronous one.

Reply:
Problem is that we can't run def_use on these methods as its currently structured, as it tends to crash with a BUG_CHECK in Definitions::getPoints because it can't figure out where the function is called from.  We need some way of constructing a ProgramPoint corresponding to the (non-existant) call site before we can analyze the body.

As far a what the annotations should be and which functions they should be on, the two are equivalent -- you're now back to something more like the original suggestion before it was later suggested they should be on the abstract methods.

Reply:
We should be able to create a ProgramPoints for asynchronous functions (which have no explicit caller). The same problem appears in several other places. There's a "beforeStart" ProgramPoint which may work.